README.md
# pureharm-db-core-jdbc

See [changelog](./CHANGELOG.md).

We do not even pretend to support anything other than Postgresql.

Utility library to be used in other pureharm modules:

- `pureharm-db-flyway`
- `pureharm-db-doobie`
- `pureharm-db-slick`
- and other modules that depend on JDBC.

## Scala versions

- Scala `2.13`: JVM
- Scala `3`: JVM

## modules

- `"com.busymachines" %% s"pureharm-db-core-jdbc" % "0.6.0"`. Which has these as its main dependencies:
  - [pgjdbc](https://github.com/pgjdbc/pgjdbc/releases) `42.2.23`
  - [pureharm-core-anomaly](https://github.com/busymachines/pureharm-core/releases) `0.3.0`
  - [pureharm-db-core](https://github.com/busymachines/pureharm-db-core/releases) `0.5.0`

## usage

Under construction. See [release notes](https://github.com/busymachines/pureharm-db-core-jdbc/releases) and tests for examples.

## Copyright and License

All code is available to you under the Apache 2.0 license, available
at [http://www.apache.org/licenses/LICENSE-2.0](http://www.apache.org/licenses/LICENSE-2.0) and also in
the [LICENSE](./LICENSE) file.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

# unreleased

# 0.6.0

This is the first release that is also available for Scala 3!

### dependency upgrades

- [pureharm-core-anomaly](https://github.com/busymachines/pureharm-core/releases) `0.3.0`
- [pureharm-db-core](https://github.com/busymachines/pureharm-db-core/releases) `0.5.0`
- [pgjdbc](https://github.com/pgjdbc/pgjdbc/releases) `42.2.23`

### internals
- bump scalafmt to `3.0.0-RC6` — from `2.7.5`
- bump sbt to `1.5.5`
- bump sbt-spiewak to `0.21.0`


# 0.5.0

Identical to `0.6.0` for Scala 2, but it failed to release for Scala 3. Otherwise it is identical.

# 0.4.0

### dependency upgrades

- [pureharm-db-core](https://github.com/busymachines/pureharm-db-core/releases) `0.4.0`

# 0.3.0

### dependency upgrades

- [pureharm-core-anomaly](https://github.com/busymachines/pureharm-core/releases) `0.2.0`
- [pureharm-db-core](https://github.com/busymachines/pureharm-db-core/releases) `0.3.0`

# 0.2.0

- remove dependency on `pureharm-effect-cats` in `Compile`, now only in `Test`

### dependency upgrades

- bump pureharm to `0.2.0`
- bump log4cats to `1.2.2`
- bump atto to `0.9.3`

# 0.1.0

Split out from [pureharm](https://github.com/busymachines/pureharm) as of version `0.0.7`.

:warning: Breaking changes :warning:

- renamed module maven artifact ID from `pureharm-db-core-psql` to `pureharm-db-core-jdbc`
- renamed package `busymachines.pureharm.db.psql` -> `busymachines.pureharm.db.pgjdbc`


db-core-jdbc/src/test/scala/busymachines/pureharm/db/psql/PSQLExceptionParserTests.scala
/*
 * Copyright 2019 BusyMachines
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package busymachines.pureharm.db.psql

import busymachines.pureharm.effects._
import busymachines.pureharm.testkit._
import org.typelevel.log4cats.slf4j._

/** @author Lorand Szakacs, https://github.com/lorandszakacs
  * @since 02 Jul 2020
  */
final class PSQLExceptionParserTests extends PureharmTest {

  private val parsers = PSQLExceptionInterpreters.PSQLErrorParsers
  override val testLogger: TestLogger =  TestLogger(Slf4jLogger.getLogger[IO])

  test("unique") {
    for {
      attempt <- parsers.unique[IO]("Key (id)=(row1_id) already exists.").attempt
    } yield assertSuccess(attempt)(("id", "row1_id"))
  }

  test("foreign") {
    for {
      attempt <-
        parsers.foreignKey[IO]("""Key (row_id)=(120-3921-039213) is not present in table "pureharm_rows".""").attempt
    } yield assertSuccess(attempt)(("row_id", "120-3921-039213", "pureharm_rows"))
  }
}


db-core-jdbc/src/test/scala/busymachines/pureharm/db/pgjdbc/PSQLExceptionParserTests.scala
/*
 * Copyright 2019 BusyMachines
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package busymachines.pureharm.db.pgjdbc

import cats.effect.IO
import busymachines.pureharm.testkit._
import org.typelevel.log4cats.slf4j._

/** @author
  *   Lorand Szakacs, https://github.com/lorandszakacs
  * @since 02
  *   Jul 2020
  */
final class PSQLExceptionParserTests extends PureharmTest {

  private val parsers = PSQLExceptionInterpreters.PSQLErrorParsers
  override val testLogger: TestLogger = TestLogger(Slf4jLogger.getLogger[IO])

  test("unique") {
    for {
      attempt <- parsers.unique[IO]("Key (id)=(row1_id) already exists.").attempt
    } yield assertSuccess(attempt)(("id", "row1_id"))
  }

  test("foreign") {
    for {
      attempt <-
        parsers.foreignKey[IO]("""Key (row_id)=(120-3921-039213) is not present in table "pureharm_rows".""").attempt
    } yield assertSuccess(attempt)(("row_id", "120-3921-039213", "pureharm_rows"))
  }
}


db-core-jdbc/src/main/scala/busymachines/pureharm/db/pgjdbc/PSQLExceptionDebug.scala
/*
 * Copyright 2019 BusyMachines
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package busymachines.pureharm.db.pgjdbc

import cats.Show
import org.postgresql.util.PSQLException

object PSQLExceptionDebug {

  def apply(e: PSQLException): String = PSQLExceptionShow.show(e)

  implicit val PSQLExceptionShow: Show[PSQLException] = Show.show { e =>
    val msg = e.getServerErrorMessage
    s"""
       |
       |---------------------
       |class    = ${e.getClass.getCanonicalName}
       |toString = ${e.toString}
       |
       |msg parts:
       |
       |state      = ${e.getSQLState}
       |column     = ${msg.getColumn}
       |datatype   = ${msg.getDatatype}
       |detail     = ${msg.getDetail}
       |file       = ${msg.getFile}
       |hint       = ${msg.getHint}
       |line       = ${msg.getLine}
       |message    = ${msg.getMessage}
       |position   = ${msg.getPosition}
       |routine    = ${msg.getRoutine}
       |schema     = ${msg.getSchema}
       |severity   = ${msg.getSeverity}
       |table      = ${msg.getTable}
       |where      = ${msg.getWhere}
       |constraint = ${msg.getConstraint}
       |internal_query    = ${msg.getInternalQuery}
       |internal_position = ${msg.getInternalPosition}
       |""".stripMargin
  }
}


db-core-jdbc/src/main/scala/busymachines/pureharm/db/pgjdbc/PSQLExceptionInterpreters.scala
/*
 * Copyright 2019 BusyMachines
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package busymachines.pureharm.db.pgjdbc

import cats._
import cats.syntax.all._
import busymachines.pureharm.db._
import org.postgresql.util._

/** @author
  *   Lorand Szakacs, https://github.com/lorandszakacs
  * @since 02
  *   Jul 2020
  */
object PSQLExceptionInterpreters {

  object PSQLStates {
    val NotNullViolation:    String = PSQLState.NOT_NULL_VIOLATION.getState
    val ForeignKeyViolation: String = PSQLState.FOREIGN_KEY_VIOLATION.getState
    val UniqueViolation:     String = PSQLState.UNIQUE_VIOLATION.getState
  }

  private[pgjdbc] object PSQLErrorParsers {
    import atto._, Atto._
    private val underscore  = char('_')
    private val closeParens = char(')')
    private val doubleQuote = char('"')
    private val column: Parser[String] = many1(letterOrDigit | underscore).map(_.mkString_(""))

    private val untilNextDoubleQuote: Parser[String] =
      manyUntil(anyChar, doubleQuote).map(_.mkString(""))

    /** Consumes the last parenthesis as well
      */
    private def untilLastClosedParens: Parser[String] = for {
      head      <- manyUntil(anyChar, closeParens).map(_.mkString(""))
      remaining <- get
      tail      <- if (remaining.contains(')')) untilLastClosedParens else "".pure[Parser]
    } yield s"$head$tail"

    /** usually has the value of format:
      * {{{
      *   Key (id)=(row1_id) already exists.
      * }}}
      */
    private[pgjdbc] object unique {

      private val parser: Parser[(String, String)] = for {
        _          <- string("Key (")
        columnName <- column
        _          <- string(")=(")
        value      <- untilLastClosedParens
        _          <- string(" already exists.")
      } yield (columnName, value)

      def apply[F[_]: MonadThrow](s: String): F[(String, String)] =
        parser.parse(s).done.either.leftMap(s => new RuntimeException(s): Throwable).liftTo[F]

    }

    /** example:
      * {{{
      *   Key (row_id)=(120-3921-039213) is not present in table "pureharm_rows".
      * }}}
      */
    private[pgjdbc] object foreignKey {

      private val parser: Parser[(String, String, String)] = for {
        _          <- string("Key (")
        columnName <- column
        _          <- string(")=(")
        value      <- untilLastClosedParens
        _          <- untilNextDoubleQuote
        table      <- untilNextDoubleQuote
      } yield (columnName, value, table)

      def apply[F[_]: MonadThrow](s: String): F[(String, String, String)] =
        parser.parse(s).done.either.leftMap(s => new RuntimeException(s): Throwable).liftTo[F]
    }

  }

  /** Only call when org.postgresql.util.PSQLException#getSQLState == org.postgresql.util.PSQLStates.UniqueViolation]]
    *
    * Will attempt to extract the values of the state by doing regex over the error message... yey, java?
    */
  def uniqueKey[F[_]: MonadThrow](e: PSQLException): F[DBUniqueConstraintViolationAnomaly] =
    PSQLExceptionInterpreters.PSQLErrorParsers
      .unique[F](e.getServerErrorMessage.getDetail)
      .map(t => DBUniqueConstraintViolationAnomaly(t._1, t._2))

  /** Only call when org.postgresql.util.PSQLException#getSQLState == org.postgresql.util.PSQLStates.ForeignKeyViolation
    *
    * Will attempt to extract the values of the state by doing regex over the error message... yey, java?
    */
  def foreignKey[F[_]: MonadThrow](e: PSQLException): F[DBForeignKeyConstraintViolationAnomaly] = {
    val msg = e.getServerErrorMessage
    for {
      table      <- Option(msg.getTable).liftTo[F](e)
      constraint <- Option(msg.getConstraint).liftTo[F](e)
      t          <- PSQLErrorParsers.foreignKey[F](msg.getDetail)
      (column, value, fTable) = t
    } yield DBForeignKeyConstraintViolationAnomaly(
      table        = table,
      constraint   = constraint,
      column       = column,
      value        = value,
      foreignTable = fTable,
    )
  }

  lazy val adapt: PartialFunction[Throwable, Throwable] = {
    case e: PSQLException =>
      e.getSQLState match {
        case PSQLStates.UniqueViolation     => uniqueKey[Either[Throwable, *]](e).getOrElse(e)
        case PSQLStates.ForeignKeyViolation => foreignKey[Either[Throwable, *]](e).getOrElse(e)
        case _                              => e
      }
    case e => e
  }
}
